File: sort/bogo_sort/python/bogo_sort.py
from random import randrange


def bogosort(lst):
    # Helper that returns a shuffled version of the list
    def shuffle(l):
        out = [None for x in range(len(l))]
        for item in l:
            r = randrange(0, len(l))
            while out[r] is not None:
                r -= 1
            out[r] = item
        return out

    # Helper that returns true if each element is smaller or equal than the previous
    def isinorder(l):
        for index in range(1,len(l)):
            if l[index-1] > l[index]:
                return False
        return True

    # Shuffle until in order
    while not isinorder(lst):
        lst = shuffle(lst)
        print("lst: ", lst)  # Uncomment to print lists tried

    # Finally in order
    return lst

def test():
	list = [14,37,22,84,2,31]
	sorted_list = bogosort(list)
	assert (sorted_list == sorted(list)) == True

if __name__ == "__main__":
    test()
    print(bogosort([1, 2, 3]))
    print(bogosort([3, 2, 1]))
    print(bogosort([2, 3, 1]))
    print(bogosort([2, 6, 1, 5, 7, 4, 3]))
    print(bogosort(['c', 'a', 'b']))
    print(bogosort([randrange(0,13) for i in range(13)]))  # Will take a long time (probably)

File: sort/bogo_sort/C/bogosort.c
#include<stdio.h>
#include<stdlib.h>

int sorted (int *array, int size) {
    while (size >= 1) {
        if (array[size] < array[size-1]) return 0;
        size--;
    }
    return 1;
}

void shuffle (int *array, int size) {
    int i, tmp, random;
    for (i = 0; i < size; i++) {
        tmp = array[i];
        random = rand()%size;
        array[i] = array[random];
        array[random] = tmp;
    }
}

void bogosort (int *array, int size) {
    while (!sorted(array, size)) shuffle(array, size);
}

void print_array (int *array, int size) {
    int i;
    for (i = 0; i < size; i++) {
        printf("%d ", array[i]);
    }
    printf("\n");
}

int main() {
    int i, size;
    printf("Enter array size: ");
    scanf("%d", &size);
    int array[size];
    for (i = 0; i < size; i++) {
        array[i] = rand()%100;
    }
    printf("Original array: \n");
    print_array(array, size);
    bogosort(array, size);
    printf("Sorted array: \n");
    print_array(array, size);
    return 0;
}